File: backend/routers/goals.py
"""Goals router (Feature #5) — savings / FIRE targets with progress."""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from core.db import get_db
from core.deps import get_current_user
from models.user import User
from schemas.goal import GoalCreate, GoalOut
from services import goals as goals_service

router = APIRouter(prefix="/goals", tags=["goals"])


@router.get("", response_model=list[GoalOut])
async def list_goals(
    current: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> list[GoalOut]:
    rows = await goals_service.list_goals_with_progress(db, current.id)
    return [GoalOut.model_validate(r) for r in rows]


@router.post("", response_model=GoalOut, status_code=status.HTTP_201_CREATED)
async def create_goal(
    body: GoalCreate,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> GoalOut:
    await goals_service.create_goal(
        db, current.id, body.title, body.target_usd, body.target_date
    )
    rows = await goals_service.list_goals_with_progress(db, current.id)
    # Return the newly created goal (last by created_at).
    return GoalOut.model_validate(rows[-1])


@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response)
async def delete_goal(
    goal_id: uuid.UUID,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Response:
    ok = await goals_service.delete_goal(db, current.id, goal_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Цель не найдена")
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/services/goals.ts
import { api } from "./api";

export interface Goal {
  id: string;
  title: string;
  target_usd: string;
  target_date: string | null;
  current_usd: string;
  remaining_usd: string;
  progress_pct: string | null;
  achieved: boolean;
  monthly_growth_usd: string | null;
  projected_date: string | null;
  created_at: string | null;
}

export async function listGoals(): Promise<Goal[]> {
  const { data } = await api.get<Goal[]>("/goals");
  return data;
}

export async function createGoal(body: {
  title: string;
  target_usd: number;
  target_date?: string | null;
}): Promise<Goal> {
  const { data } = await api.post<Goal>("/goals", body);
  return data;
}

export async function deleteGoal(id: string): Promise<void> {
  await api.delete(`/goals/${id}`);
}
